# Bài tập 6

Bùi Văn Hợp - 1712046

## Câu 1 (1 điểm)

Ta có Deterministic Noise như định nghĩa nằm ở bản chất của từng tập Hypothesis có khả năng xấp xỉ tốt target, khá tương tự như phần bias trong phân tích Bias-Variance. Với cách suy nghĩ trên, nếu ta sử dụng 1 tập $H' \subset H$, ta đang cố giới hạn lại khả năng của tập $H'$ so với tập $H$, bằng một cách nào đó, có thể là Regularization.

Điều đó đồng nghĩa với việc khả năng một tập $H'$ đó bị overfitting giảm lại vì Stochastic Noise (nếu tập $H$ bị overfitting do vì chúng quá fit vào những phần lỗi Stochastic đó), nhưng ngược lại khả năng chúng có Deterministic Noise lại cao hơn, vì khả năng fit của $H'$ đã bị giới hạn và giảm xuống đi so với $H$, vậy nên Deterministic Noise của $H'$ sẽ cao hơn $H$.

**Vậy mình sẽ chọn đáp án là b, In general, deterministic noise will increase.**

YOUR ANSWER HERE b

## Câu 2 (1 điểm)

In [1]:
import numpy as np

In [2]:
def get_train_data(file_name):
    train_data = np.loadtxt(file_name)
    X_train = train_data[:, :2]
    Y_train = train_data[:, -1]
    return X_train, Y_train

def get_test_data(file_name):
    test_data = np.loadtxt(file_name)
    X_test = test_data[:, :2]
    Y_test = test_data[:, -1]
    return X_test, Y_test

In [3]:
X_train, Y_train = get_train_data('in.dta')
X_test, Y_test = get_test_data('out.dta')

In [4]:
def non_linear_transform(X):
    N = X.shape[0]
    X = np.hstack((np.ones((N, 1)), X)) 
    x1 = X[:, 1].reshape(N, 1)
    x2 = X[:, 2].reshape(N, 1)
    X = np.hstack((X, x1**2))
    X = np.hstack((X, x2**2))
    X = np.hstack((X, x1 * x2))
    X = np.hstack((X, np.abs(x1 - x2)))
    X = np.hstack((X, np.abs(x1 + x2)))
    return X
    

In [5]:
def run_linear_regression(X, Y, lamd = 0):
    XX = np.dot(X.T, X)
    n = XX.shape[0]
    X_dagger = np.dot(np.linalg.pinv(XX + lamd * np.eye(n)), X.T)
    w = np.dot(X_dagger, Y)
    return w

In [6]:
def get_fraction_error(X, Y, w):
    error = np.mean(np.sign(np.dot(X, w)) != np.sign(Y))
    return error

In [7]:
# Linear Regression with Regularization and Non-linear Transformation
#Lambda = 0.0 (No Regularization)
X_train_transform = non_linear_transform(X_train)
X_test_transform = non_linear_transform(X_test)
lamd = 0.0
w = run_linear_regression(X_train_transform, Y_train, lamd = lamd)
E_in = get_fraction_error(X_train_transform, Y_train, w)
E_out = get_fraction_error(X_test_transform, Y_test, w)
print('With lambda = ', lamd)
print('Error_in : ', E_in)
print('Error out:', E_out)

With lambda =  0.0
Error_in :  0.02857142857142857
Error out: 0.084


Với bài tập Linear Regression này, kết quả cho $E_{in} = 0.02857142857142857$ và $E_{out} = 0.084$. **Mình chọn đáp án gần nhất với 2 kết quả lỗi này là a, 0.03, 0.08.**

YOUR ANSWER HERE a

## Câu 3 (1 điểm)

In [8]:
# Linear Regression with Regularization and Non-linear Transformation
# Lambda = 10^-3
X_train_transform = non_linear_transform(X_train)
X_test_transform = non_linear_transform(X_test)
lamd = 1e-3
w = run_linear_regression(X_train_transform, Y_train, lamd = lamd)
E_in = get_fraction_error(X_train_transform, Y_train, w)
E_out = get_fraction_error(X_test_transform, Y_test, w)
print('With lambda = ', lamd)
print('Error_in : ', E_in)
print('Error out:', E_out)

With lambda =  0.001
Error_in :  0.02857142857142857
Error out: 0.08


Với câu 3, ta thêm Regularization cho Linear Regression với $\lambda = 10^{-3}$. Ta có kết quả độ lỗi lần lượt cho $E_{in} và E_{out}$ là 0.02857142857142857, 0.08. **Mình sẽ lựa chọn đạp án gần nhất là d.**

YOUR ANSWER HERE d

## Câu 4 (1 điểm)

In [9]:
# Linear Regression with Regularization and Non-linear Transformation
# Lambda = 10^3
X_train_transform = non_linear_transform(X_train)
X_test_transform = non_linear_transform(X_test)
lamd = 1e3
w = run_linear_regression(X_train_transform, Y_train, lamd = lamd)
E_in = get_fraction_error(X_train_transform, Y_train, w)
E_out = get_fraction_error(X_test_transform, Y_test, w)
print('With lambda = ', lamd)
print('Error_in : ', E_in)
print('Error out:', E_out)

With lambda =  1000.0
Error_in :  0.37142857142857144
Error out: 0.436


Với câu 4, ta thêm Regularization với $\lambda = 10^{3}$, ta có kết quả độ lỗi cho $E_{in}, E_{out}$ lần lượt là 0.37142857142857144, 0.436. **Mình sẽ chọn đáp án gần nhất là e, 0.4, 0.4**

YOUR ANSWER HERE e

## Câu 5 (1 điểm)

In [10]:
best_error_out = 1.0
best_K = None
X_train_transform = non_linear_transform(X_train)
X_test_transform = non_linear_transform(X_test)
for k in range(-10, 11):
    lamd = pow(10, k)
    w = run_linear_regression(X_train_transform, Y_train, lamd = lamd)
    E_in = get_fraction_error(X_train_transform, Y_train, w)
    E_out = get_fraction_error(X_test_transform, Y_test, w)
    if E_out < best_error_out:
        best_error_out = E_out
        best_K = k
print('Best K :', best_K)
print('Best error_out : ', best_error_out)

Best K : -1
Best error_out :  0.056


Ở đây chúng ta duyệt qua lần lượt các số K (biểu diễn cho giá trị lambda) để thêm Regularization cho bài toán Linear Regression. Ở đây K = -1 cho ta độ lỗi tốt nhất là 0.056. **Với câu 5, mình sẽ chọn là đáp án d, -1.**

YOUR ANSWER HERE d

## Câu 6 (1 điểm)

Độ lỗi tốt nhất khi thêm Regularization là 0.056 từ đoạn code ở câu 5. **Nên mình chọn đáp án gần nhất là b, 0.06.**

YOUR ANSWER HERE b

## Câu 7 (1 điểm)

Ta sẽ kiểm tra cho từng đáp án.

Ta có 1 nhận xét từ các biểu thức ràng buộc các hypothesis là :
1. $H(Q, 0, Q_0) = H(Q_0 - 1)$, đây là một ràng buộc cho các tập Hypothesis bằng cách cho các bậc cao hơn có hệ sống bằng 0. 

2. $H(0) \subset H(1) \subset H(2) ... H(Q)$

Với [a], $H(10, 0, 3) \cup H(10, 0, 4) = H(2) \cup H(3) = H(3)$. Vậy câu a sai.

Với [b], $H(10, 1, 3) \cup H(10, 1, 4) \neq H(3)$. Đây là điều khá dễ thấy từ nhận xét ở trên.

Với [c], $H(10, 0, 3) \cap H(10, 0, 4) = H(2) \cap H(3) = H(2)$. Với câu c thì ta nhận được 1 đáp án đúng.

Với [d], $H(10, 1, 3) \cap H(10, 1, 4) \neq H(1)$. Đây là điều hiển nhiên tương tự như câu b.

Vậy đáp án có biểu thức đúng là c.

YOUR ANSWER HERE c

## Câu 8 (1 điểm)

Ta có một fully connected Neural Network, với số lớp L = 2, $d^{(0)} = 5, d^{(1)} = 3, d^{(2)} = 1$. Ta sẽ đếm lần lượt các phép tính là: $w_{ij}^{(l)}x_i^{(l-1)} , w_{ij}^{(l)}\delta_j^{(l)}, x_i^{(l - 1)}\delta_j^{(l)}$

Với phép tính $w_{ij}^{(l)}x_i^{(l-1)}$, ta có các phép tính này xuất hiện trong phần Lan truyền tiến (Forward-propagation): 

- Từ Lớp 0 đi qua Lớp 1 có (5 + 1) * 3 = 18 phép tính.

- Từ Lớp 1 đi qua Lớp 2 có (3 + 1) * 1 = 4 phép tính. 

Vậy tổng cộng có **22 phép tính** $w_{ij}^{(l)}x_i^{(l-1)}$.

Với phép tính $w_{ij}^{(l)}\delta_j^{(l)}$, ta có các phép tính này xuất hiện trong phần Lan truyền ngược (Backpropagation):

- Với Lớp 2, ta không cần phải tính $\delta_0^{(2)}$ sử dụng phép tính trên, vì chúng ta sẽ tính đạo hàm đó từ độ lỗi cần tối ưu.

- Ta chỉ cần sử dụng phép tính $w_{ij}^{(l)}\delta_j^{(l)}$ trong lớp thứ 1, vì lớp 0 là lớp input không cần tính. Thêm nữa ta chỉ cần tính cho 3 node, trừ node 0 của Lớp 2 không có phép tính này, nên ta chỉ cần sử dụng **3 phép tính** $w_{ij}^{(l)}\delta_j^{(l)}$.

Với phép tính $x_i^{(l - 1)}\delta_j^{(l)}$, chúng ta sử dụng chúng trong việc cập nhật lại các tham số $w_{ij}$ ở từng lớp, nên tương tự như với phép tính $w_{ij}^{(l)}x_i^{(l-1)}$, ta cũng sử dụng **22 phép tính**.

**Vậy tổng cộng ta sử dụng 47 phép tính. Mình chọn đáp án gần nhất là 45, d**

YOUR ANSWER HERE d

## Câu 9 (1 điểm)

Ta có một input gồm 10 unit (kể cả $x_0^{(0)}$) và một output với 1 unit. Ta có 36 hidden unit. 

Số lượng unit trong mạng nơ-ron này có số lượng tham số (weight) ít nhất là khi ta xếp mỗi lớp chỉ có một unit, ta có 36 lớp ẩn chỉ có 1 unit.

**Vậy số lượng unit trong mạng này là: 10 * 1 + (1 * 1 * 35) + 1 * 1 = 46** (1 * 1 * 35 vì giữa 36 lớp thì 35 weighted matrix chứa các $w_{ij}$). Nên đáp án mình chọn là a, 46.

YOUR ANSWER HERE a

## Câu 10 (1 điểm)

In [11]:
#With only 1 hidden layer with 36 hidden units in. We have 10 * 35 + 36 * 1 = 386 weights w_ij
n_hidden = 36
max_w_size = 386
u1_max = None
u2_max = None
u3_max = None
# Test with 2 hidden layers
for u1 in range(1, n_hidden + 1):
    for u2 in range(1, n_hidden + 1):
        if u1 + u2 == n_hidden:
            w_size = 10 * (u1 - 1) + u1 * (u2 - 1) + u2
            if max_w_size < w_size:
                max_w_size = w_size
                u1_max = u1
                u2_max = u2
            
# Test with 3 hidden layers
for u1 in range(1, n_hidden + 1):
    for u2 in range(1, n_hidden + 1):
        for u3 in range(1, n_hidden + 1):
            if u1 + u2 + u3 == n_hidden:
                w_size = 10 * (u1 - 1) + u1 * (u2 - 1) + u2 * (u3 - 1) + u3
                if max_w_size < w_size:
                    max_w_size = w_size
                    u1_max = u1
                    u2_max = u2
                    u3_max = u3
                    
print("maximum size of weighted w_ij", max_w_size)
print("Neural Network structure:")
if u1_max != None: 
    print("u1 : ", u1_max)
if u2_max != None: 
    print("u2 : ", u2_max)
if u3_max != None: 
    print("u3 : ", u3_max)

maximum size of weighted w_ij 510
Neural Network structure:
u1 :  22
u2 :  14


Mình chạy đoạn code trên để thử các trường hợp có thể có của mạng với 2 Lớp hoặc 3 Lớp ẩn cho 36 unit ẩn trên. Mình có được 1 cấu trúc mạng cho ra số weight lớn nhất là $d^{(1)} = 22, d^{(2)} = 14$ với số weight đếm được là 510. **Nên đáp án của mình là e, 510.**

YOUR ANSWER HERE e